streams/src/main/java/org/apache/kafka/streams/processor/internals/AssignedStreamsTasks.java

```diff
@@ -49,48 +52,233 @@ public StreamTask restoringTaskFor(final TopicPartition partition) {
+
+        for (final TaskId task : revokedTasks) {
+            if (running.containsKey(task)) {
+                revokedRunningTasks.add(task);
+            } else if (created.containsKey(task)) {
+                revokedNonRunningTasks.add(task);
+            } else if (restoring.containsKey(task)) {
+                revokedRestoringTasks.add(task);
+            } else {
+                log.error("Task {} was revoked but cannot be found in the assignment", task);
+                firstException.set(new IllegalStateException("Revoked a task that was not in the assignment"));
+            }
+        }
+
+        firstException.compareAndSet(null, suspendRunningTasks(revokedRunningTasks, revokedChangelogs));
+        firstException.compareAndSet(null, closeNonRunningTasks(revokedNonRunningTasks, revokedChangelogs));
+        firstException.compareAndSet(null, closeRestoringTasks(revokedRestoringTasks, revokedChangelogs));
+
+        return firstException.get();
+    }
+
+    private RuntimeException suspendRunningTasks(final Set<TaskId> revokedRunningTasks, final List<TopicPartition> revokedChangelogs) {
+        final AtomicReference<RuntimeException> firstException = new AtomicReference<>(null);
+        for (final Iterator<TaskId> it = revokedRunningTasks.iterator(); it.hasNext(); ) {
+            final TaskId id = it.next();
+            final StreamTask task = running.get(id);
+
+            // if the task is already suspended just log a message and move on
+            if (suspended.containsKey(id)) {
+                log.error("Found already suspended task {} while suspending tasks", task);
+            } else {
+                try {
+                    task.suspend();
+                    suspended.put(id, task);
+                } catch (final TaskMigratedException closeAsZombieAndSwallow) {
+                    // as we suspend a task, we are either shutting down or rebalancing, thus, we swallow and move on
+                    log.info("Failed to suspend {} {} since it got migrated to another thread already. " +
+                        "Closing it as zombie and move on.", taskTypeName, id);
+                    firstException.compareAndSet(null, closeZombieTask(task));
+                } catch (final RuntimeException e) {
+                    log.error("Suspending {} {} failed due to the following error:", taskTypeName, id, e);
+                    firstException.compareAndSet(null, e);
+                    try {
+                        task.close(false, false);
+                    } catch (final RuntimeException f) {
+                        log.error(
+                            "After suspending failed, closing the same {} {} failed again due to the following error:",
+                            taskTypeName, id, f);
+                    }
+                } finally {
+                    it.remove();
+
+                    // Wait to reset changelog partitions and remove from running/runningByPartition until the task is
```

Comment:
Now I understand that suspended is refactored to be just a subset of running.

But for the changelog topics, I'd still suggest we still remove them since in `StateStoreChangelogReader`:

```
startRestoration(final Set<TopicPartition> initialized,
                                  final RestoringTasks active) {
        log.debug("Start restoring state stores from changelog topics {}", initialized);

        final Set<TopicPartition> assignment = new HashSet<>(restoreConsumer.assignment());
        assignment.addAll(initialized);
        restoreConsumer.assign(assignment);
```

We always incrementally add the new assignment anyways, so this is not really saving us much.

Reply:
Only running tasks are suspended though, so they should not be restored if resumed. It doesn't matter for the `restoreConsumer` since completed changelogs are removed in `updateNewAndRestoringTasks`, but the `storeChangelogReader` does keep track of restored offsets. 
We should just clear these from the ChangelogReader iff they are closed, because they wouldn't be added back to it if just resumed

Reply:
SG.

Reply:
Yeah that's what I've read from the code.